### PR TITLE
Revert "Revert "Logout endpoint should use POST not GET""

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

Original PR #1830 
Reverts transcom/mymove#1858 

Two things fixed in this PR:

1) "Sign Out" link now acts like a real link for mouse rollover
2) Login.gov redirect is moved from the server to the client

## More detail

The redirect was a bit complicated because of the use of `fetch()`. Unfortunately, while `fetch()` will follow redirects from the server the redirect will use pre-flight checks against CORS redirects that can't be turned off.  This is despite the fact that a 303 should allow a POST request to be redirected to a GET request.  If `fetch()` allowed you to intercept the redirect to manually walk the redirect chain then we would do that, but the `redirect: 'manual'` doesn't do this, it instead returns an opaque response with no actionable information.  

The only reasonable thing to do then is to process the post request with a 200 response and pass back the logout URL for Login.gov. The client then has to do the redirect itself by using `window.location.href` which acts like the GET request we'd want in the first place.  This does perform as expected and is probably the simplest way forward here. The consequence of not handling this URL is that the user remains logged into Login.gov and on a subsequent visit to `/auth/login` would re-authenticate and get a new session token.

The alternative is to not let the client have any say and use a `<form>` instead of an `<a onClick=...>`, but that opens up a world where we'd need the form to be styled as though it was a link to not change the behavior requirements from design.  However, using a form would allow the server to do all of the work of logout AND redirect.

Question for reviewers: Is this introducing a security flaw? Or are we confident that letting the client log itself out is sufficient for our application?

## Testing

This needs to be tested in three ways:

- [x] Dev environment devlocal login/logout works
- [x] Dev environment login.gov login/logout works
- [x] Experimental login.gov login/logout works

### CODE REVIEW

File: src/shared/User/api.js [hunk 1]
```diff
@@ -1,8 +1,28 @@
...
+  const req = {
+    url: logoutEndpoint,
+    method: 'POST',
+    credentials: 'same-origin', // Passes through CSRF cookies
+    requestInterceptor,
+  };
+  try {
+    // Successful logout should return a redirect url
+    let resp = await Swagger.http(req);
+    window.location.href = resp.text;
```
Comment (chrisgilmerproj) [inline src/shared/User/api.js]: After reading this article I think this is the best choice: https://appendto.com/2016/04/javascript-redirect-how-to-redirect-a-web-page-with-javascript/

File: pkg/auth/authentication/auth.go [hunk 2]
```diff
@@ -113,17 +113,18 @@ func (h LogoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+			// This operation will delete all cookies from the session
 			session.IDToken = ""
 			session.UserID = uuid.Nil
 			auth.WriteSessionCookie(w, session, h.clientAuthSecretKey, h.noSessionTimeout, h.logger, h.useSecureCookie)
-			http.Redirect(w, r, logoutURL, http.StatusTemporaryRedirect)
+			fmt.Fprint(w, logoutURL)
 		} else {
 			// Can't log out of login.gov without a token, redirect and let them re-auth
-			http.Redirect(w, r, redirectURL, http.StatusTemporaryRedirect)
+			fmt.Fprint(w, redirectURL)
```
Comment (chrisgilmerproj) [inline pkg/auth/authentication/auth.go]: This is where we pass back the Logout URL in the response.

Should this be formatted as a JSON object?

File: src/shared/User/api.js (hunk 1 above)
Comment (donaldthai) [reply src/shared/User/api.js]: I have an alternative solution to client side redirection. If we can follow the redirection via form post, then we can just set the request header `'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8'` to mimic a form post without the html form tag!

More info here:
- https://github.com/github/fetch/issues/263#issuecomment-279968330
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST

Comment (chrisgilmerproj) [reply src/shared/User/api.js]: Ok, this seemed very promising so I tried it.  I run into the same CORS problem as I did before, so perhaps I was wrong that a POST from a form would make this work.  Here's the diff:

```diff
diff --git a/pkg/auth/authentication/auth.go b/pkg/auth/authentication/auth.go
index b5dbeb575..ab7c4762a 100644
--- a/pkg/auth/authentication/auth.go
+++ b/pkg/auth/authentication/auth.go
@@ -121,10 +121,10 @@ func (h LogoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
                        session.IDToken = ""
                        session.UserID = uuid.Nil
                        auth.WriteSessionCookie(w, session, h.clientAuthSecretKey, h.noSessionTimeout, h.logger, h.useSecureCookie)
-                       fmt.Fprint(w, logoutURL)
+                       http.Redirect(w, r, logoutURL, http.StatusSeeOther)
                } else {
                        // Can't log out of login.gov without a token, redirect and let them re-auth
-                       fmt.Fprint(w, redirectURL)
+                       http.Redirect(w, r, redirectURL, http.StatusSeeOther)
                }
        }
 }
diff --git a/src/shared/User/api.js b/src/shared/User/api.js
index 83c1d0957..9f469b7d0 100644
--- a/src/shared/User/api.js
+++ b/src/shared/User/api.js
@@ -11,18 +11,22 @@ export async function GetLoggedInUser() {

 export async function LogoutUser() {
   const logoutEndpoint = '/auth/logout';
+  const headers = {
+    'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
+  };
   const req = {
     url: logoutEndpoint,
     method: 'POST',
     credentials: 'same-origin', // Passes through CSRF cookies
+    headers,
     requestInterceptor,
   };
   try {
     // Successful logout should return a redirect url
     let resp = await Swagger.http(req);
-    window.location.href = resp.text;
+    // window.location.href = resp.text;
   } catch (err) {
     // Failure to logout should return user to homepage
-    window.location.href = '/';
+    // window.location.href = '/';
   }
 }
```

Comment (donaldthai) [reply src/shared/User/api.js]: That's too bad. It was worth a shot.

File: pkg/auth/authentication/auth.go (hunk 2 above)
Comment (donaldthai) [reply pkg/auth/authentication/auth.go]: Seems fine to me like this.

**Pull request merged**